insights/parsers/lspci.py

```diff
@@ -103,4 +111,91 @@ def pci_dev_list(self):
         """
         The list of PCI devices.
         """
-        return self.data.keys()
+        return list(self.data.keys())
+
+
+@parser(Specs.lspci_vmmkn)
+class LsPciVmmkn(CommandParser, list):
+    """
+    Class to parse the PCI device information gathered from the
+    ``/sbin/lspci -vmmkn`` command.
+
+    Typical output of the ``lspci -vmmkn`` command is::
+
+        Slot:   00:00.0
+        Class:  0600
+        Vendor: 8086
+        Device: 1237
+        SVendor:    1af4
+        SDevice:    1100
+        Rev:    02
+
+        Slot:   00:01.0
+        Class:  0101
+        Vendor: 8086
+        Device: 7010
+        SVendor:    1af4
+        SDevice:    1100
+        ProgIf: 80
+        Driver: ata_piix
+        Module: ata_piix
+        Module: ata_generic
+
+        Slot:   00:01.1
+        Class:  0c03
+        Vendor: 8086
+        Device: 7020
+        SVendor:    1af4
+        SDevice:    1100
+        Rev:    01
+        Driver: uhci_hcd
+
+        Slot:   00:03.0
+        Class:  0200
+        Vendor: 1af4
+        Device: 1000
+        SVendor:    1af4
+        SDevice:    0001
+        PhySlot:    3
+        Driver: virtio-pci
+
+    Examples:
+        >>> type(lspci_vmmkn)
+        <class 'insights.parsers.lspci.LsPciVmmkn'>
+        >>> sorted(lspci_vmmkn.pci_dev_list)
+        ['00:00.0', '00:01.0', '00:01.1', '00:03.0']
+        >>> lspci_vmmkn[0].get('Driver') is None
+        True
+        >>> lspci_vmmkn[-1].get('Driver')
+        'virtio-pci'
+
+    Attributes:
+
+    """
+    def parse_content(self, content):
```

Comment:
Per my test on some hosts, the items of each HW device sometimes are different due to the type of the device. I'm afraid that the `re.findall` may not work well for these kinds of dynamic items. Please let me know your details if I understand incorrectly.

Reply:
Take a look here https://asciinema.org/a/9TL6DD5tvceD2kcIC7acrKCFj. I did some simple cli app for parsing. The main parsing logic is here: https://gitlab.com/zhukovgreen/lspci-parser/-/blob/master/temp_example_of_lspci_parsing/app.py#L24

Reply:
Your example looks good, however, it depends on a set of hard-coded `"container_keys"` and only parses these keys. The insights parser is trying to parse all the content and make them serialized in case there are other rules that require other `keys`.

And I believe the `re.findall` would work w/o the hard-coded `keys` as well. But besides that, we still need to `split` the whole content according to the `Solt` (now the blank lines). And at last, the code may look at the same complex level.